### PR TITLE
Add guidelines for making F# data types structs

### PR DESCRIPTION
**Pull request by cartermp**
This was inspired from this SO question: https://stackoverflow.com/questions/59738472/struct-attribute-on-discriminated-unions

### CODE REVIEW

File: docs/fsharp/style-guide/conventions.md [hunk 1]
```diff
@@ -437,11 +437,120 @@ Finally, automatic generalization is not always a boon for people who are new to
...
 ## Performance
 
+### Prefer structs for small data types
+
+Using structs (also called Value Types) can often result in higher performance for some code because it avoids allocating and de-allocating objects on the managed heap in .NET. However, structs are not always a "go faster" button: if the size of the data in a struct exceeds 16 bytes, copying the data can often result in more CPU time spend than using a reference type.
+
+A good rule of thumb for using structs intead of value types is:
+
+* If the size of your data is 16 bytes or smaller (measure with a call to `sizeof`)
+* If your data type is made up of other value types (e.g., a record of `int`s)
```
Comment (Youssef1313) [inline docs/fsharp/style-guide/conventions.md]:
```suggestion
* If your data type is made up of other value types (for example, a record of `int`s)
```

https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg

File: docs/fsharp/style-guide/conventions.md [hunk 2]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
 
 ## Performance
 
+### Prefer structs for small data types
+
+Using structs (also called Value Types) can often result in higher performance for some code because it typically avoids allocating objects. However, structs are not always a "go faster" button: if the size of the data in a struct exceeds 16 bytes, copying the data can often result in more CPU time spend than using a reference type.
+
+To determine if you should use a struct, consider the following two points:
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]: not sure if you need to number them

```suggestion
To determine if you should use a struct, consider the following conditions:
```

File: docs/fsharp/style-guide/conventions.md [hunk 3]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
+
+1. If the size of your data is 16 bytes or smaller
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]:
```suggestion
- If the size of your data is 16 bytes or smaller.
```

File: docs/fsharp/style-guide/conventions.md [hunk 4]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
+2. If you are likely to have many of these data types resident in memory in a running program
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]:
```suggestion
- If you're likely to have many of these data types resident in memory in a running program.
```

File: docs/fsharp/style-guide/conventions.md [hunk 5]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
+
+If the first point applies, you should generally use a struct. If both apply, you should almost always use a struct.
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]:
```suggestion
If the first condition applies, you should generally use a struct. If both apply, you should almost always use a struct.
```

File: docs/fsharp/style-guide/conventions.md [hunk 6]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
+
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]: Should you explain why almost always and not always?

File: docs/fsharp/style-guide/conventions.md [hunk 7]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
+        t
+    else
+        let struct(x, y, z) = t
+        let r = offsetValues x y z offset
+        runWithStructTuple r offset (times - 1)
+```
+
+When you benchmark these functions with a statistical benchmarking tool like [BenchmarkDotNet](https://benchmarkdotnet.org/), you'll find that the `runWithStructTuple` function that uses struct tuples runs 40% faster and allocates no memory.
+
+However, these results will not always be the case in your own code. Depeneding on if you mark a function as `inline`, code that uses reference tuples may get some additional optimizations, or code that would allocate could simply be optimized away. You should always measure results whenever performance is concerned, and never operate based on assumption or intuition.
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]:
```suggestion
However, these results won't always be the case in your own code. If you mark a function as `inline`, code that uses reference tuples may get some additional optimizations, or code that would allocate could simply be optimized away. You should always measure results whenever performance is concerned, and never operate based on assumption or intuition.
```

File: docs/fsharp/style-guide/conventions.md [hunk 8]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
+        { p with X = p.X + offset; Y = p.Y + offset; Z = p.Z + offset }
+
+    if times <= 0 then
+        p
+    else
+        let r = offsetValues p offset
+        processStructPoint r offset (times - 1)
+```
+
+This is very similar to the tuple code before, but this time using records and an inlined inner function.
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]:
```suggestion
This is very similar to the previous tuple code, but this time the example uses records and an inlined inner function.
```

File: docs/fsharp/style-guide/conventions.md [hunk 9]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
+        |> Name
+
+    let structReverseName (SName s) =
+        s.ToCharArray()
+        |> Array.rev
+        |> string
+        |> SName
+```
+
+It's very common to define single-case Discriminated Unions like this for domain modeling. When you benchmark these functions with a statistical benchmarking tool like [BenchmarkDotNet](https://benchmarkdotnet.org/) you'll find that `structReverseName` runs about 25% faster than `reverseName` for small strings. For very large strings, both will perform about the same, so in this case it is always preferable to use a struct. As mentioned before, always measure and do not operate on assumptions or intuition.
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]:
```suggestion
It's very common to define single-case Discriminated Unions like this for domain modeling. When you benchmark these functions with a statistical benchmarking tool like [BenchmarkDotNet](https://benchmarkdotnet.org/), you'll find that `structReverseName` runs about 25% faster than `reverseName` for small strings. For very large strings, both perform about the same. So, in this case, it's always preferable to use a struct. As previously mentioned, always measure and do not operate on assumptions or intuition.
```

File: docs/fsharp/style-guide/conventions.md [hunk 10]
```diff
@@ -437,11 +437,118 @@ Finally, automatic generalization is not always a boon for people who are new to
...
 
-However, use of `mutable` in F# may feel at odds with functional purity. This is fine, if you adjust expectations from purity to [referential transparency](https://en.wikipedia.org/wiki/Referential_transparency). Referential transparency - not purity - is the end goal when writing F# functions. This allows you to write a functional interface over a mutation-based implementation for performance critical code.
+Use of `mutable` in F# may feel at odds with functional purity. This is understandable, but functional purity everywhere can be at odds with performance goals. A compromise is to encapsulate mutation such that callers need not care about what happens when they call a funciton. This allows you to write a functional interface over a mutation-based implementation for performance critical code.
```
Comment (mairaw) [inline docs/fsharp/style-guide/conventions.md]: F# has some funny names so not sure if funciton is one of them :-) 

```suggestion
Use of `mutable` in F# may feel at odds with functional purity. This is understandable, but functional purity everywhere can be at odds with performance goals. A compromise is to encapsulate mutation such that callers need not care about what happens when they call a function. This allows you to write a functional interface over a mutation-based implementation for performance-critical code.
```

**Review by mairaw — APPROVED** [review]
LGTM. Left a few comments. 

You might also consider checking acrolinx: https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/014b4251-abb4-407a-882f-1125554f2433
Ideally, the minimum across for each category would be 80.

File: docs/fsharp/style-guide/conventions.md (hunk 6 above)
Comment (cartermp) [reply docs/fsharp/style-guide/conventions.md]: Yep, did. Gist of it is that if you measure, you could still find no difference or perhaps structs being worse. So it's always important to measure.

**Pull request merged**